### PR TITLE
Numpy 2 update

### LINKED ISSUE #37840: Migrate to Numpy v2.0
Maintenance for moving to numpy v2.0 (need issue links) ([migration guide](https://numpy.org/doc/stable/numpy_2_0_migration_guide.html))

- [ ] Need prototype(s) for how extensive the changes might be
- [ ] Will be done on a long-running branch to integrate the changes. `main` will be auto-merged into the branch
- [x] numpy v2 supports python 3.9-3.12 ([python supported versions](https://devguide.python.org/versions/))
- [x] boost v1.84 supports newer numpy [boost::python docs](https://www.boost.org/doc/libs/1_84_0/libs/python/doc/html/numpy/index.html) https://github.com/mantidproject/mantid/pull/37766
- [x] Move to matplotlib >=3.8.4 https://github.com/mantidproject/mantid/pull/38077/

Mantid will not move to matplotlib=3.9 until later because it requires qt=6

### PR DESCRIPTION
**Pull request by jhaigh0**
### Description of work
Update to using Numpy 2.

 - Added the `NPY201` ruff rule which automatically updated a lot of breaking api changes
 - Updated code which resulted in test failures
 - Update Pystog to v0.5 which supports numpy 2

There was a matplotlib compatibility issue in one place, see fc7cd5d791c277b3994fa1a2232905cec67b4c77 for details.

Fixes #37840

### To test:

 - https://builds.mantidproject.org/job/build_packages_from_branch/953/
 - After the package build you can test it install from the `numpy_2_test` label
 - Should do some light manual testing

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

### CODE REVIEW

**Review by jclarkeSTFC — DISMISSED** [review]
Installed the standalone and conda package on IDAaaS, both worked fine with some light testing, code looks good.

File: docs/source/release/v6.12.0/Framework/Dependencies/New_features/38420.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+- Updated Numpy to version 2. `Read about the changes <https://numpy.org/news/#numpy-200-released>`_. **Users should note that Numpy 2 introduces some breaking API changes, see the `Numpy 2 Migration Guide <https://numpy.org/devdocs/numpy_2_0_migration_guide.html>`_ for more details**
```
Comment (thomashampson) [inline docs/source/release/v6.12.0/Framework/Dependencies/New_features/38420.rst:1]: Normally we are specific about minor version support when updating our depenency on numpy.
```suggestion
- Drop support for NumPy version 1. We now build against NumPy v2.0 and support up to v2.1. `Read about the changes <https://numpy.org/news/#numpy-200-released>`_. **Users should note that NumPy 2 introduces some breaking API changes. See the `NumPy 2 Migration Guide <https://numpy.org/devdocs/numpy_2_0_migration_guide.html>`_ for more details**
```

**Review by thomashampson — CHANGES_REQUESTED** [review]
Changes all seem relatively straightforward. The update to the Conda recipes look fine. I just have a minor suggestion to make sure the release notes are specific enough.

**Review by jclarkeSTFC — APPROVED** [review]

**Review by thomashampson — APPROVED** [review]
Thanks for the release notes change. All looking good now.

**Pull request merged**